README.txt
           LibSBMLSim: The library for simulating SBML models

                     LibSBMLSim development team
             http://fun.bio.keio.ac.jp/software/libsbmlsim/
                  mailto:sbmlsim@fun.bio.keio.ac.jp

-- Last modified: Tue, 05 Dec 2017 00:12:42 +0900

* Overview
  LibSBMLSim is a library for simulating an SBML model which contains
  Ordinary Differential Equations (ODEs). LibSBMLSim provides simple
  command-line tool and several APIs to load an SBML model, perform
  numerical integration (simulate) and export its results.
  Both explicit and implicit methods are supported on libSBMLSim.
  LibSBMLSim is confirmed to pass all SBML Level-2 Version 4 and Level-3
  Version 1 core test cases (sbml-test-cases-2014-10-22.zip, available from
  http://sourceforge.net/projects/sbml/files/test-suite/3.1.1/).
  The libSBMLSim code is portable. It is written in C programming language
  (ANSI C89) and it does not depend on other third-party libraries
  except libSBML(*1).
  The library should build and work without serious troubles on Unix
  based operating systems (Linux, MacOSX and FreeBSD) and on Windows
  (with Visual C++).
  LibSBMLSim also provides several language bindings like Java, Python,
  C# and Ruby. Perl binding is already included in the source tree, but
  is not able to create through the single build process (see the
  description below).

  (*1 libSBML: http://sbml.org/Software/libSBML)

  LibSBMLSim can be used to create your own SBML capable simulator,
  plug-in, web based application and web services. The API is quite
  straight forward. You can run a simulation and generate a result
  file in Comma Separated Values (CSV) with a few lines of codes.
  === Python ============================
    from libsbmlsim import *
    r = simulateSBMLFromFile('sbml.xml', 20.0, 0.1, 10, 0, MTHD_RUNGE_KUTTA, 0)
    write_csv(r, 'result.csv')
  =======================================

  Please see the 'API.txt' and 'examples' directory for further information
  on libSBMLSim APIs.

* Installation
- Dependencies
  LibSBMLSim requires libSBML to be installed on your system.
  Please follow the instruction on (*1) and install libSBML.
  LibSBML and its dependent libraries will be automatically installed
  with Windows version of libSBMLSim Installer.

- Binary install of libSBMLSim
  We have provided installer for Windows, MacOSX and Linux from libSBMLSim-1.3.
  = Windows (both 32bit and 64bit)
    Download libSBMLSim Installer for Windows (libsbmlsim-1.4.0-win{32,64}.exe)
    and double-click the installer. It will ask few questions, and will
    install libSBMLSim to
      "C:\Program Files\libsbmlsim-1.3"       (64bit)
      "C:\Program Files (x86)\libsbmlsim-1.3" (32bit)
    by default.

  = MacOSX (64bit)
    Download libSBMLSim Installer archive for MacOSX
    (libsbmlsim-1.4.0-macosx-mavericks-x64.dmg) and double-click the .dmg file.
    You will see an installer (libsbmlsim-1.4.0-macosx-mavericks-x64.pkg) in a
    Finder window. Double-click the installer and follow the instructions.
    It will install libSBMLSim to /usr/local .

  = Linux (64bit)
    Download libsbmlsim-1.4.0_amd64.deb, and install with the following
    command in your terminal.
      $ sudo dpkg -i libsbmlsim-1.4.0_amd64.deb
    It will install libSBMLSim to /usr .

* Compile and Install from source code.
- Required software packages to compile libSBMLSim
  CMake(*2) is required to compile libSBMLSim. Please download
  and install CMake-2.8.12 or above from (*2) before building libSBMLSim.
  If you want to use language bindings of libSBMLSim, please
  download and install SWIG-2.0.4 or above from (*3).
  (Note: If you installed SWIG from MacPorts, please install
         swig-java, swig-python, swig-ruby, swig-csharp which are required
         to compile language bindings for libSBMLSim.)

  (*2 CMake: http://cmake.org/)
  (*3 SWIG:  http://swig.org/)

- How to build libSBMLSim
  1. Extract the archive file
   % tar xvzf libsbmlsim-1.4.0.tar.gz (for tar ball)
   % unzip libsbmlsim-1.4.0.zip       (for zip archive)
  2. Compile
   % mkdir libsbmlsim-1.4.0/build
   % cd libsbmlsim-1.4.0/build
   % cmake ..
   % ccmake .
     CUI from cmake will be launched. Please confirm that
     cmake have automatically detected the installed location of
     libSBML. You can check the installed location from the
     following values:
     (ex. on MacOSX)
       LIBSBML_INCLUDE_DIR            /usr/local/include
       LIBSBML_LIBRARY                /usr/local/lib/libsbml.dylib

     If libSBML is not detected automatically, you can manually
     specify the installed location through this menu.

     If you want to build language bindings, please turn on the
     corresponding compile option.
       WITH_JAVA     ... build with Java bindings
       WITH_PYTHON   ... build with Python bindings
       WITH_RUBY     ... build with Ruby bindings
       WITH_CSHARP   ... build with C# bindings

     Once you press [c] key, cmake will run the configure procedure
     and tries to detect SWIG, Java, Python, C# and Ruby (depending on
     which language bindings you enabled). Hit [c] several times to
     complete configuration. Once the configuration is done,
     press [g] key and cmake will generate Makefile.
     After Makefile is generated, just run

   % make
   % sudo make install
     which will compile and install the library, command-line tool
     and header files on your system. Default prefix (install
     directory) is
       - /usr/local                  ... on Linux and MacOSX
       - C:\Program Files\libsbmlsim ... on Windows
     (Note: You can change the prefix from the UI of ccmake)
     Also, you can create binary installer by following command.
   % make package

- Installed files
  Following files are installed on your system.
  = Unix based systems (Linux, MacOSX, etc.)
    $prefix/bin/simulateSBML         ... SBML simulator
    $prefix/lib/libsbmlsim-static.a  ... Static library
               /libsbmlsim.dylib     ... Dynamic library (on MacOSX)
               /libsbmlsim.so        ... Dynamic library (on Linux)
    $prefix/include/libsbmlsim       ... Header files
    $prefix/share/libsbmlsim/        ... Sample files (SBML, results)
                            /c       ... Sample C code
                            /cpp     ... Sample C++ code
                            /csharp  ... Sample C# code and language bindings
                            /java    ... Sample Java code and language bingings
                            /python  ... Sample Python code and language bingings
                            /ruby    ... Sample Ruby code and language bingings

  = Windows
    $prefix\bin\simulateSBML.exe     ... SBML simulator
    $prefix\bin\sbmlsim.dll          ... Dynamic library
    $prefix\lib\sbmlsim-static.lib   ... Static library
               \sbmlsim.lib          ... Lib file for dynamic library
    $prefix\include\libsbmlsim       ... Header files
    $prefix\share\libsbmlsim\        ... Sample files (SBML, results)
                            \c       ... Sample C code
                            \cpp     ... Sample C++ code
                            \csharp  ... Sample C# code and language bindings
                            \java    ... Sample Java code and language bingings
                            \python  ... Sample Python code and language bingings
    (Note: Ruby binding is not supported on Windows)

* Usage
- simulateSBML
  simulateSBML is a simple SBML simulator which accept SBML file as
  an input, and then output "out.csv" as a result.
  Usage: simulateSBML [option] filename(SBML)
    -t #    : specify simulation time (ex. -t 100 )
    -s #    : specify simulation step (ex. -s 100 )
    -d #    : specify simulation delta (ex. -d 0.01 [default:1/4096])
              dt is calculated in (delta)*(time)/(step)
    -a      : print Species Value in Amount
    -o file : specify result file (ex. -o output.csv )
    -l      : use lazy method for integration
    -n      : do not use lazy method
    -v      : prints version info
    -A #    : specify absolute tolerance for variable stepsize (ex. -A 1e-03 [default:1e-09])
    -R #    : specify relative tolerance for variable stepsize (ex. -R 0.1   [default:1e-06])
    -M #    : specify the max change rate of stepsize (ex. -M 1.5 [default:2.0])
    -B      : use bifurcation analysis
    -m #    : specify numerical integration algorithm (ex. -m 3 )
           1: Runge-Kutta
           2: AM1 & BD1 (implicit Euler)
           3: AM2 (Crank Nicolson)
           4: AM3
           5: AM4
           6: BD2
           7: BD3
           8: BD4
           9: AB1 (explicit Euler)
          10: AB2
          11: AB3
          12: AB4
          13: Runge-Kutta-Fehlberg
          14: Cash-Karp
          (AM: Adams-Moulton, BD: Backward-Difference, AB: Adams-Bashforth.
           Number after synonim specifies the order of integration.
           For example, AM2 is "2nd order Adams-Moulton" method)

- Scripts for "SBML test cases"
  LibSBMLSim provides scripts to easily run SBML test cases (*4)
  and compare the results with it. Generated results are compatible
  with Online SBML Test Suite (*4), so you can run all tests with
  this scripts and upload the results to Online SBML Test Suite.
  The scripts are not installed, you will find them under "testcases"
  directory in the extracted source directory (libsbmlsim/testcases/).

    libsbmlsim/testcases/simulateSBML  ... SBML simulator
                        /runall.sh     ... Script which will run all tests
                        /compare.pl
                        /genresult.pl
                        /wrapper.sh    ... Wrapper script for SBML Test Runner

  "simulateSBML" simulates SBML model and generates simulation result
  as a CSV file, which is identical with the one installed under
  $prefix/bin . "runall.sh" will call simulateSBML for all SBML test
  cases, and compare the result with the one from SBML test cases.
  "compare.pl" and "genresult.pl" are scripts which will support some
  functions called from runall.sh.
  The SBML test cases are not included in this distribution, so please
  download them from (*5). After downloading sbml-test-cases-X.Y.Z.zip,
  unzip the archive and move (or copy) "cases/" directory to
  libsbmlsim/testcases directory. The directory structure will be:

    libsbmlsim/testcases/simulateSBML  ... SBML simulator
                        /runall.sh     ... Script which will run all tests
                        /compare.pl
                        /genresult.pl
                        /wrapper.sh    ... Wrapper script for SBML Test Runner
                        /cases/semantic/00001 ... Test case 1
                        /cases/semantic/00002 ... Test case 2
                        /cases/semantic/00003 ... Test case 3
                        /cases/semantic/...

  Following command will test all 1,125 tests and print out the
  results, whether the simulation result matches with the result
  with the one from SBML test cases.

    % ./runall.sh
    00001: 5 : 50 : [S1,S2] : [S1,S2] : [S1,S2] : 1e-16 : 1e-10
      print amount
      time:5 step:50 dt:0.100000
      Model 00001 ... [OK]
    00002: 5.0 : 50 : [S1,S2] : [S1,S2] : [S1,S2] : 1e-16 : 1e-10
      print amount
      time:5 step:50 dt:0.100000
      Model 00002 ... [OK]
    00003: 5.0 : 50 : [S1,S2] : [S1,S2] : [S1,S2] : 1e-16 : 1e-10
      print amount
      time:5 step:50 dt:0.100000
      Model 00003 ... [OK]

  If the simulation result doesn't match with the one from
  SBML test cases, then the result will be marked as "[NG]".

  (*4 Online SBML Test Suite: http://sbml.org/Facilities/Online_SBML_Test_Suite)
  (*5 SBML-test-cases-3.1.1 : http://sourceforge.net/projects/sbml/files/test-suite/3.1.1)

  Note: Because 00983/00983-sbml-l3v1.xml is an invalid SBML document in
        SBML-test-cases-3.1.1, libSBMLSim will not run a simulation for this
        model. We confirmed that libSBMLSim will pass the test of
        00983/00983-sbml-l3v1.xml from SBML-test-cases-3.2.0.

- Wrapper script for "SBML Test Runner"
  As from v1.3.0, libSBMLSim provides a wrapper script, "wrapper.sh" for
  SBML Test Runner (*6). By using wrapper.sh, uesrs can easily run all
  SBML test cases on libSBMLSim through a Graphical User Interface of
  SBML Test Runner.
  To run libSBMLSim through SBML Test Runner, you have to create a
  configuration for libSBMLSim from [Preference] menu on SBML Test Runner.
  In the "Preferences" dialog, please assign a name to the configuration
  (ex. libsbmlsim), and fill out the text fields as follows:
    = Name: libsbmlsim
    = Wrapper path: $some_where/libsbmlsim/testcases/wrapper.sh
    = Output directory: $some_where/output
    = Unsupported tags: comp, fbc
    = Arguments to wrapper: %d %n %o %l %v
    you can turn on "Wrapper can handle any SBML Level/Version" and
    "Wrapper can be run in parallel".
  After filling out the text fields, save the configuration and then
  selecting [Test] -> [Run All Supported Tests] from the menu, SBML
  Test Runner will run all test cases. You can browse its results on
  its GUI.
  (*6 SBML Test Runner: https://github.com/sbmlteam/sbml-test-suite)

* LibSBMLSim API and its language bindings
  Example usage of libSBMLSim APIs are as follows.
  Please see the 'API.txt' and 'examples' directory for further information
  on libSBMLSim APIs.

- C, C++ API
  === C code ============================
  #include "libsbmlsim/libsbmlsim.h"
  ...
  /*
   * Simulate sbml.xml to time=20 with dt=0.1, print_interval=10
   * by 4th-order Runge-Kutta Method.
   */
  myResult *r = simulateSBMLFromFile("sbml.xml", 20, 0.1, 10, 0, MTHD_RUNGE_KUTTA, 0);
  /*
   * Export simulation result as CSV file
   */
  write_csv(r, "result.csv");
  /*
   * Free Result object
   */
  free_myResult(r);
  =====================================

- Java, Python, Ruby bindings
  LibSBMLSim API is also provided for several language bindings.
  === Python ============================
  from libsbmlsim import *
  r = simulateSBMLFromFile('sbml.xml', 20.0, 0.1, 10, 0, MTHD_RUNGE_KUTTA, 0)
  write_csv(r, 'result.csv')
  =======================================

  === Java ==============================
  import jp.ac.keio.bio.fun.libsbmlsim.*;
  ...
  System.loadLibrary("sbmlsimj");
  myResult r = libsbmlsim.simulateSBMLFromFile("sbml.xml", 20.0, 0.1, 10, 0, libsbmlsim.MTHD_RUNGE_KUTTA, 0);
  libsbmlsim.write_csv(r, "result.csv");
  =======================================

  === Ruby ==============================
  require 'libsbmlsim'
  r = Libsbmlsim::simulateSBMLFromFile('sbml.xml', 20.0, 0.1, 10, 0, Libsbmlsim::MTHD_RUNGE_KUTTA, 0)
  Libsbmlsim::write_csv(r, 'result.csv')
  =======================================

  === C# ================================
  using System;
  public class Test
  {
    static void Main()
      {
        myResult result = libsbmlsim.simulateSBMLFromFile("sbml.xml", 20.0, 0.1, 10, 0, libsbmlsim.MTHD_RUNGE_KUTTA, 0);
        libsbmlsim.write_csv(result, "test.csv");
      }
  }
  =======================================

  Please see the 'API.txt' and 'examples' directory for further information.
  The 'examples' directory contains sample code for test application
  in several programming languages (C, C++, Java, Python, Ruby, C# and Perl).

Have fun!
--
LibSBMLSim development team <sbmlsim@fun.bio.keio.ac.jp>


NEWS.txt
           libSBMLSim NEWS -- History of user-visible changes

- Last modified: Tue, 05 Dec 2017 01:07:05 +0900

======================================================================
Version 1.4.0 (2017-12-05)
======================================================================

* New features

  - New memory management system was included to avoid memory leak.

* Bug fixes

  - Several bug fixes on memory leak. Now, LibSBMLSim has small memory leak
    on limited number of models (which include reversible fast reactions).

  - Add support on CMake 3.x so that it can generate Java8 bindings.

  - Fixed crash on python bindings which causes double free on gc.

  libsbmlsim confirmed to pass SBML test cases from 00001 to 01185.

======================================================================
Version 1.3.0 (2016-12-23)
======================================================================

* New features

  - Add wrapper script (wrapper.sh) for SBML Test Runner.

  - Add -o option so that user can specify an output file. This will
    enable SBML Test Runner to run wrapper.sh in parallel.

* Bug fixes

  - Allocate including_species dynamically, so that the library can
    handle more than 256 species.

  - Fixed warnings (CMP0037 CMP0042) from cmake.

  - Fixed alter_tree_structure to call ASTNode_setName() in the
    proper order.

  - Removed unused function (mySpecies_setLocatingCompartment()) from
    mySpecies.

  - Add -lm to linker option.

  - Avoid a C preprocessor warning on Linux.

  libsbmlsim confirmed to pass SBML test cases from 00001 to 01185.

======================================================================
Version 1.2.0 (2013-04-04)
======================================================================

* New features

  - CMake uses C_INCLUDE_PATH, DYLD_LIBRARY_PATH and
    LD_LIBRARY_PATH to find libSBML.

* Bug fixes

  - Fixed to support fast and reversible reaction
    (00986, 00987, 00988).

  - Fixed crash when ASTNode only have 1 or 0 child under '+', '*',
    'AND', 'OR', 'XOR' ASTNodes. (01112, 01113, 01114)

  - Fixed crash when piecewise have odd number of children (which
    means when piecewise doesn't have "otherwise" node)
    (01115, 01116).

  Now, libsbmlsim passed all the tests from 00001 to 01185.

======================================================================
Version 1.1.0 (2013-01-08)
======================================================================

* New features

  - New integrators (5th order Runge-Kutta Fehlberg, 5th order Cash-Karp)
    are included in this release. Both new integrators support variable
    step size integration.
     = 5th order Runge-Kutta Fehlberg : MTHD_RUNGE_KUTTA_FEHLBERG_5
     = 5th order Cash-Karp : MTHD_NAME_CASH_KARP

  - The C# bindings functions are include in this release. Please see
    examples/csharp/Test.cs for the sample code.
    
  - Both 32bit(x86) and 64bit(x86_64) build for Windows are available.
  
  - Binary Installer is available for Windows, MacOSX and Linux.

* Bug fixes

  - Memory leaks in free_mySBML_objects() have been fixed.

* API changes

  - myResult* simulateSBMLModel() now requires 3 additional parameters for
    variable step integration. 
    myResult* simulateSBMLModel(Model_t *m, double sim_time, double dt,
                                int print_interval, int print_amount,
                                int method, int use_lazy_method,
                                double atol, double rtol, double facmax);
     = double atol;  Absolute error tolernace
     = double rtol;  Relative error tolerance
     = double facmax; Maximum acceptable increasing factor
    If you don't know the exact value for above 3 values, please just assign
    "0.0" for all 3 values, then libsbmlsim will use the default values for
    the integration.

======================================================================
Version 1.0.0 (2012-04-04)
======================================================================

* Initial release, 04 April 2012.



src/main.c
/*
 * <!--------------------------------------------------------------------------
 * This file is part of libSBMLSim.  Please visit
 * http://fun.bio.keio.ac.jp/software/libsbmlsim/ for more
 * information about libSBMLSim and its latest version.
 *
 * Copyright (C) 2011-2017 by the Keio University, Yokohama, Japan
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation.  A copy of the license agreement is provided
 * in the file named "LICENSE.txt" included with this software distribution.
 * ---------------------------------------------------------------------- -->*/
#include "libsbmlsim/libsbmlsim.h"

#if defined(_MSC_VER) || defined(__STRICT_ANSI__)
#include "libsbmlsim/my_getopt.h"
#else
#include <unistd.h>
#endif

void usage(char *str) {
  printf("Usage    : %s [option] filename(SBML file only)\n", str);
  printf(" -t #    : specify simulation time (ex. -t 100 )\n");
  printf(" -s #    : specify simulation step (ex. -s 100 )\n");
  printf(" -d #    : specify simulation delta (ex. -d 0.01 [default:1/4096])\n");
  printf("           dt is calculated in (delta)*(time)/(step)\n");
  printf(" -a      : print Species Value in Amount\n");
  printf(" -o file : specify result file (ex. -o output.csv )\n");
  printf(" -l      : use lazy method for integration\n");
  printf(" -n      : do not use lazy method\n");
  printf(" -v      : prints version info\n");
  printf(" -A #    : specify absolute tolerance for variable stepsize (ex. -A 1e-03 [default:1e-09])\n");
  printf(" -R #    : specify relative tolerance for variable stepsize (ex. -R 0.1   [default:1e-06])\n");
  printf(" -M #    : specify the max change rate of stepsize (ex. -M 1.5 [default:2.0])\n");
  printf(" -B      : use bifurcation analysis\n");
  printf(" -m #    : specify numerical integration algorithm (ex. -m 3 )\n");
  printf("        1: Runge-Kutta\n");
  printf("        2: AM1 & BD1 (implicit Euler)\n");
  printf("        3: AM2 (Crank Nicolson)\n");
  printf("        4: AM3\n");
  printf("        5: AM4\n");
  printf("        6: BD2\n");
  printf("        7: BD3\n");
  printf("        8: BD4\n");
  printf("        9: AB1 (explicit Euler)\n");
  printf("       10: AB2\n");
  printf("       11: AB3\n");
  printf("       12: AB4\n");
  printf("       13: Runge-Kutta-Fehlberg\n");
  printf("       14: Cash-Karp\n");
  exit(1);
}

int main(int argc, char *argv[]){
  SBMLDocument_t *d;
  Model_t *m;
  /*  Variables for getopt() */
  int ch;
  extern char *optarg;
  extern int optind;
  unsigned int error_num;

  char *myname;
  boolean use_lazy_method = 2;
  int print_amount = 0;

  double sim_time = 0;
  int step = 0;
  double delta = 1.0/4096;
  double dt = 0;
  int print_interval = 0;
  char* outfile = NULL;

  double atol = ABSOLUTE_ERROR_TOLERANCE;
  double rtol = RELATIVE_ERROR_TOLERANCE;

  char buf1[256], buf2[256], buf3[256];
  char *tmp;
  int method;
  boolean is_explicit;
  char *method_name;
  int method_key = -1;

  /*for bifurcation analysis*/
  boolean use_bifurcation_analysis = false;

  /*for variable step-size numerical integration*/
  boolean use_variable_stepsize = false;
  double facmax = DEFAULT_FACMAX;

  myResult *rtn;

  myname = argv[0];
  while ((ch = getopt(argc, argv, "t:s:d:m:A:R:M:o:lnaBv")) != -1){
    switch (ch) {
      case 't':
        sim_time = atof(optarg);
        break;
      case 's':
        step = atoi(optarg);
        break;
      case 'd':
        delta = atof(optarg);
        break;
      case 'm':
        method_key = atoi(optarg);
        break;
      case 'l':
        use_lazy_method = true;
        break;
      case 'n':
        use_lazy_method = false;
        break;
      case 'v':
        printf("%s version %s\n", myname, LIBSBMLSIM_DOTTED_VERSION);
        exit(0);
      case 'a':
        print_amount = 1;
        break;
      case 'A':
        atol = atof(optarg);
        break;
      case 'R':
        rtol = atof(optarg);
        break;
      case 'B':
        use_bifurcation_analysis = 1;
        break;
      case 'M':
        facmax = atof(optarg);
        break;
      case 'o':
        outfile = optarg;
        break;
      default:
        usage(myname);
    }
  }
  if (outfile == NULL) {
    outfile = "out.csv";
  }
  argc -= optind;
  argv += optind;

  /* get SBML Model */
  if(argc < 1){
    usage(myname);
  }
  d = readSBML(argv[0]);

  error_num = SBMLDocument_getNumErrors(d);
  if (error_num > 0) {
    unsigned int i;
    for (i = 0; i < error_num; i++) {
      const SBMLError_t *err = SBMLDocument_getError(d, i);
      if (XMLError_isError((XMLError_t *)err) || XMLError_isFatal((XMLError_t *)err)) {
        printf("Input file [%s] is not an appropriate SBML file\n", argv[0]);
        exit(1);
      }
    }
  }
  m = SBMLDocument_getModel(d);

  /* determine sim_time */
  if(sim_time == 0){
    while(1){
      printf("Simulation time : ");
      tmp = fgets(buf1, 256, stdin);
      chomp(buf1);
      if(str_is_number(buf1)){
        break;
      }
      printf("not a number!\n");
    }
    sscanf(buf1, "%lf", &sim_time);
  }
  /* determine step */
  if(step == 0){
    while(1){
      printf("simulation step : ");
      tmp = fgets(buf1, 256, stdin);
      chomp(buf1);
      if(str_is_number(buf1)){
        break;
      }
      printf("not a number!\n");
    }
    sscanf(buf1, "%d", &step);
  }

  /* CUI */
  if (method_key == -1) {
    while(1){
      printf("select neumerical integration method\n");
      printf("simulate with\n");
      printf("Runge-Kutta : press \"1\"\n");
      printf("AM1 & BD1 (implicit Euler) : press \"2\"\n");
      printf("AM2 (Crank Nicolson) : press \"3\"\n");
      printf("AM3 : press \"4\"\n");
      printf("AM4 : press \"5\"\n");
      printf("BD2 : press \"6\"\n");
      printf("BD3 : press \"7\"\n");
      printf("BD4 : press \"8\"\n");
      printf("AB1 (explicit Euler) : press \"9\"\n");
      printf("AB2 : press \"10\"\n");
      printf("AB3 : press \"11\"\n");
      printf("AB4 : press \"12\"\n");
	  printf("RKF : press \"13\"\n");
	  printf("CK  : press \"14\"\n");
      tmp = fgets(buf2, 256, stdin);
      method_key = atoi(buf2);
      if (method_key < 1 || method_key > 14) {
        printf("Invalid Input!\nSelect and input the number \"1~12\"");
      } else {
        break;
      }
    }
  }
  switch(method_key) {
    case 1: /*  Runge-Kutta */
      method = MTHD_RUNGE_KUTTA;
      method_name = MTHD_NAME_RUNGE_KUTTA;
      break;
    case 2: /*  Backward-Euler */
      method = MTHD_BACKWARD_EULER;
      method_name = MTHD_NAME_BACKWARD_EULER;
      break;
    case 3: /*  Crank-Nicolson */
      method = MTHD_CRANK_NICOLSON;
      method_name = MTHD_NAME_CRANK_NICOLSON;
      break;
    case 4: /*  Adams-Moulton 3 */
      method = MTHD_ADAMS_MOULTON_3;
      method_name = MTHD_NAME_ADAMS_MOULTON_3;
      break;
    case 5: /*  Adams-Moultion 4 */
      method = MTHD_ADAMS_MOULTON_4;
      method_name = MTHD_NAME_ADAMS_MOULTON_4;
      break;
    case 6: /*  Backward-Difference 2 */
      method = MTHD_BACKWARD_DIFFERENCE_2;
      method_name = MTHD_NAME_BACKWARD_DIFFERENCE_2;
      break;
    case 7: /*  Backward-Difference 3 */
      method = MTHD_BACKWARD_DIFFERENCE_3;
      method_name = MTHD_NAME_BACKWARD_DIFFERENCE_3;
      break;
    case 8: /*  Backward-Difference 4 */
      method = MTHD_BACKWARD_DIFFERENCE_4;
      method_name = MTHD_NAME_BACKWARD_DIFFERENCE_4;
      break;
    case 9: /*  Euler (Adams-Bashforth) */
      method = MTHD_EULER;
      method_name = MTHD_NAME_EULER;
      break;
    case 10: /*  Adams-Bashforth 2 */
      method = MTHD_ADAMS_BASHFORTH_2;
      method_name = MTHD_NAME_ADAMS_BASHFORTH_2;
      break;
    case 11: /*  Adams-Bashforth 3 */
      method = MTHD_ADAMS_BASHFORTH_3;
      method_name = MTHD_NAME_ADAMS_BASHFORTH_3;
      break;
    case 12: /*  Adams-Bashforth 4 */
      method = MTHD_ADAMS_BASHFORTH_4;
      method_name = MTHD_NAME_ADAMS_BASHFORTH_4;
      break;
    case 13: /*  Runge-Kutta-Fehlberg 5 */
      method = MTHD_RUNGE_KUTTA_FEHLBERG_5;
      method_name = MTHD_NAME_RUNGE_KUTTA_FEHLBERG_5;
      use_variable_stepsize = 1;
      break;
    case 14: /*  Cash-Karp */
      method = MTHD_CASH_KARP;
      method_name = MTHD_NAME_CASH_KARP;
      use_variable_stepsize = 1;
      break;
    default:
      method = MTHD_RUNGE_KUTTA;
      method_name = MTHD_NAME_RUNGE_KUTTA;
      break;
  }
  is_explicit = method % 10;

  /* simulation */
  if (!use_variable_stepsize) {
  if (!is_explicit) {
    if (use_lazy_method == 2) {
      while(1){
        printf("use lazy mode?\nlazy mode:using jacobian continuously while the solutions of equations are converging in newton method.\nyes(y) or no(n)\n");
        tmp = fgets(buf3, 256, stdin);
        if(strcmp(buf3, "y\n") == 0 || strcmp(buf3, "yes\n") == 0) {
          use_lazy_method = true;
          break;
        } else if(strcmp(buf3, "n\n") == 0 || strcmp(buf3, "no\n") == 0) {
          use_lazy_method = false;
          break;
        }
        printf("Invalid Input!!!\nSelect yes(y) or no(n)\n");
      }
    }
    if(use_lazy_method == true) {
      printf("  simulate with lazy mode\n");
    }
  }
  }
  /* Run simulation */
  if (!use_variable_stepsize) {
	  /* calculate simulation condition */
	  dt = delta*(sim_time/step);
  }else{
	  /* calculate simulation condition */
	  dt = sim_time/step;
  }
  print_interval = (int)(1/delta);
  printf("  time:%g step:%d dt:%f\n", sim_time, step, dt);
  rtn = simulateSBMLModel(m, sim_time, dt, print_interval, print_amount, method, use_lazy_method, atol, rtol, facmax);

  /* display allocated memory */
  /* MEM_TRACE(); */

  /* write CSV */
  if (rtn == NULL) {
    printf("Returned result is NULL\n");
  } else {
    write_csv(rtn, outfile); /*  for SBML test suite */
    /* to be more generic simulator
       write_separate_result(rtn,
       "./simulation_results/species_result.dat",
       "./simulation_results/parameter_result.dat",
       "./simulation_results/compartment_result.dat");
       */
  }

  /* free */
  SBMLDocument_free(d);
  free_myResult(rtn);
  /* check memory leak */
  MEM_TRACE();
  return 0;
}
